core/trino-main/src/main/java/io/trino/sql/planner/LogicalPlanner.java

```diff
@@ -519,19 +520,19 @@ private RelationPlan getInsertPlan(
                 statisticsMetadata);
     }
 
-    private static Function<Expression, Expression> failIfPredicateIsNotMeet(Metadata metadata, Session session, StandardErrorCode errorCode, String errorMessage)
+    private static Function<Expression, Expression> failIfPredicateIsNotMet(Metadata metadata, Session session, ErrorCodeSupplier errorCode, String errorMessage)
     {
-        ResolvedFunction fail = metadata.resolveFunction(session, QualifiedName.of("fail"), fromTypes(INTEGER, VARCHAR));
-        return predicate -> new IfExpression(
-                predicate,
-                TRUE_LITERAL,
-                new Cast(
-                        new FunctionCall(
-                                fail.toQualifiedName(),
-                                ImmutableList.of(
-                                        new Cast(new LongLiteral(Long.toString(errorCode.toErrorCode().getCode())), toSqlType(INTEGER)),
-                                        new Cast(new StringLiteral(errorMessage), toSqlType(VARCHAR)))),
```

Comment:
These casts are needed in the IR because the type of a StringLiteral is `varchar(n)`, and the fail function takes a `varchar`. Similarly, the type of a LongLiteral can be `integer` or `bigint` depending on the value. So, you either need to keep the cast or use a type constructor (i.e., `varchar 'xxxxxx'`, `integer '123'`) via a GenericLiteral.

Reply:
There are other places that do it this way. Why does it still work without this?

Reply:
I converted to `GenericLiteral`

Reply:
Those other places are “broken”. It works by chance just because the native container type for those types and their semantics match.